crates/sound/src/audio_backends/pw_backend.rs

```diff
@@ -160,13 +202,174 @@ impl AudioBackend for PwBackend {
+        }
         Ok(())
     }
 
-    fn prepare(&self, _stream_id: u32) -> Result<()> {
+    fn prepare(&self, stream_id: u32) -> Result<()> {
+        let mut stream_states = self.stream_states.write().unwrap();
+        let mut stream_hash = self.stream_hash.write().unwrap();
         self.thread_loop.lock();
-        self.thread_loop.unlock();
+        let stream_params = self.stream_params.read().unwrap();
+        let params = stream_params[stream_id as usize].clone();
+        //dbg!("params: {}", params);
+
+        if stream_states[stream_id as usize] == VIRTIO_SND_R_PCM_SET_PARAMS {
+            let mut buff = [0; 1024];
+            let p_buff = &mut buff as *mut i32 as *mut c_void;
+
+            let mut b = spa_pod_builder {
+                data: p_buff,
+                size: buff.len() as u32,
+                _padding: 0,
+                callbacks: spa_callbacks {
+                    funcs: std::ptr::null(),
+                    data: std::ptr::null_mut(),
+                },
+                state: spa_pod_builder_state {
+                    offset: 0,
+                    flags: 0,
+                    frame: std::ptr::null_mut(),
+                },
+            };
+
+            let mut pos: [u32; 64] = [0; 64];
+
+            //todo map position to chmap
+            pos[0] = SPA_AUDIO_CHANNEL_FL;
+            pos[1] = SPA_AUDIO_CHANNEL_FR;
+            let mut info = spa_audio_info_raw {
+                format: match params.format {
+                    VIRTIO_SND_PCM_FMT_S8 => SPA_AUDIO_FORMAT_S8,
+                    VIRTIO_SND_PCM_FMT_U8 => SPA_AUDIO_FORMAT_U8,
+                    VIRTIO_SND_PCM_FMT_S16 => SPA_AUDIO_FORMAT_S16,
+                    VIRTIO_SND_PCM_FMT_U16 => SPA_AUDIO_FORMAT_U16,
+                    VIRTIO_SND_PCM_FMT_S32 => SPA_AUDIO_FORMAT_S32,
+                    VIRTIO_SND_PCM_FMT_U32 => SPA_AUDIO_FORMAT_U32,
+                    _ => SPA_AUDIO_FORMAT_UNKNOWN,
+                },
+                rate: match params.rate {
+                    VIRTIO_SND_PCM_RATE_32000 => 32000,
+                    VIRTIO_SND_PCM_RATE_44100 => 44100,
+                    VIRTIO_SND_PCM_RATE_48000 => 48000,
+                    VIRTIO_SND_PCM_RATE_64000 => 64000,
+                    _ => 44100,
+                },
+                flags: 0,
+                channels: params.channels as u32,
+                position: pos,
+            };
+            let param: *mut spa_pod = unsafe {
+                let p = spa_format_audio_raw_build(&mut b, SPA_PARAM_EnumFormat, &mut info);
+                p
+            };
+
+            let _context = &self.context;
+            let mut stream = pw::stream::Stream::<i32>::new(
+                &self.core,
+                "audio-output",
+                properties! {
+                    *pw::keys::MEDIA_TYPE => "Audio",
+                    *pw::keys::MEDIA_CATEGORY => "Playback",
+                    *pw::keys::MEDIA_ROLE => "Music",
```

Comment:
Is this correct?

Reply:
this is just metadata, I'm not pretty sure about the media_role, but I can decide to not set it(not necessary). given that we are only configuring the metadata for playback ATM, the properties metadata should be correct.